tests/cases/compiler/genericCallInferenceInConditionalTypes1.ts

```diff
@@ -0,0 +1,59 @@
+// @strict: true
+// @noEmit: true
+
+// https://github.com/microsoft/TypeScript/issues/59937
+
+type Ref<T> = {
+  current: T;
+};
+
+type FunctionComponent<P> = (props: P) => unknown;
+
+type ComponentProps<T extends FunctionComponent<any>> =
+  T extends FunctionComponent<infer P> ? P : {};
+
+type PropsWithoutRef<P> = P extends any
+  ? "ref" extends keyof P
+    ? Omit<P, "ref">
+    : P
+  : P;
+
+type ComponentPropsWithoutRef<T extends FunctionComponent<any>> =
+  PropsWithoutRef<ComponentProps<T>>;
+
+declare function forwardRef<T, P>(
+  component: (props: P, ref: Ref<T>) => unknown,
+): (props: P & { ref?: Ref<T> }) => unknown;
+
+const ComponentWithForwardRef = forwardRef(
+  <T extends FunctionComponent<any>>(
+    props: ComponentPropsWithoutRef<T>,
+    ref: Ref<HTMLElement>,
+  ) => {
+    return null;
+  },
+);
+
+type Test<T> = T extends { component?: infer Component }
+  ? Component extends FunctionComponent<any>
+    ? ComponentProps<Component>
+    : never
+  : never;
+
+// the first one here has a chance to pollute the cache
+type Result1 = ComponentProps<typeof ComponentWithForwardRef>;
+// that could be incorrectly reused by this one
+type Result2 = Test<{ component: typeof ComponentWithForwardRef }>; // no `T` leak
```

Comment:
in the process of resolving the first conditional type here a result with a conditional type's restrictive mapper was cached, then when resolving the second conditional type that cached result was reused incorrectly, failing to produce a type with its correct mapper applied to it